### PR TITLE
[ROCKETMQ-194] log appender support

### PR DESCRIPTION
**Pull request by lindzh**
As Rocketmq is widely used,A log appender is also necessary.

### CODE REVIEW

File: logappender/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# RocketMQ-LogAppender   [![Build Status](https://travis-ci.org/rocketmq/rocketmq-logappender.svg?branch=master)](https://travis-ci.org/rocketmq/rocketmq-logappender) [![Coverage Status](https://coveralls.io/repos/github/rocketmq/rocketmq-logappender/badge.svg?branch=master)](https://coveralls.io/github/rocketmq/rocketmq-logappender?branch=master)
```
Comment (vongosling) [inline logappender/README.md]: Could we remove it if  we want it into parent modue, we can reuse rocektmq existing infra.

File: logappender/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+## Introduction
+RocketMQ-LogAppender is a logging component implement of log4j,log4j2 and logback.Taking Apache RocketMQ as broker.
```
Comment (vongosling) [inline logappender/README.md]: What is "implement of " It's confused. IMO, this project is used to send log message using RocektMQ. while log4j, log4j2 and logback is just a  adapter for SL4J

File: logappender/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## Examples
+
+#### log4j example
+
+> config detail
+
+```xml
+<appender name="mqAppender1" class="org.apache.rocketmq.logappender.log4j.RocketmqLog4jAppender">
+    <param name="Tag" value="log1" />
```
Comment (vongosling) [inline logappender/README.md]: case-insensitive name ,right ?

File: logappender/README.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    <param name="NameServerAddress" value="127.0.0.1:9876"/>
+    <layout class="org.apache.log4j.PatternLayout">
+        <param name="ConversionPattern" value="%d{yyyy-MM-dd HH:mm:ss}-%p %t %c - %m%n" />
+    </layout>
+</appender>
+
+<appender name="mqAsyncAppender1" class="org.apache.log4j.AsyncAppender">
+    <param name="BufferSize" value="1024" />
+    <param name="Blocking" value="false" />
+    <appender-ref ref="mqAppender1"></appender-ref>
```
Comment (vongosling) [inline logappender/README.md]: why not use single label as above?

File: logappender/README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    <neverBlock>true</neverBlock>
+    <appender-ref ref="mqAppender1"/>
+</appender>
+<logger name="testLogger" level="debug" additivity="false">
+    <appender-ref ref="mqAsyncAppender1"/>
+</logger>
+
+```
+
+More example,like log4j2,please see Test case.
```
Comment (vongosling) [inline logappender/README.md]: Could we give the example link for user ?

File: logappender/pom.xml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        <version>4.1.0-incubating-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>rocketmq-logappender</artifactId>
+	<packaging>jar</packaging>
+    <name>rocketmq-logappender ${project.version}</name>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.test.skip>false</maven.test.skip>
```
Comment (vongosling) [inline logappender/pom.xml]: Redundant config. we can extend from parent module.

File: logappender/pom.xml [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            <artifactId>log4j-core</artifactId>
+            <version>2.7</version>
+            <optional>true</optional>
+        </dependency>
+
+
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-core</artifactId>
+            <version>1.1.2</version>
```
Comment (vongosling) [inline logappender/pom.xml]: The latest version ?

File: logappender/pom.xml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        <maven.compiler.target>1.7</maven.compiler.target>
+    </properties>
+
+    <dependencies>
+
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>1.7.7</version>
+            <optional>true</optional>
```
Comment (vongosling) [inline logappender/pom.xml]: We must consider the suitable dependency here.

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j/RocketmqLog4jAppender.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+import org.apache.log4j.spi.LoggingEvent;
+import org.apache.rocketmq.client.producer.MQProducer;
+
+/**
+ * Log4j Appender Component
+ */
+public class RocketmqLog4jAppender extends AppenderSkeleton {
+
+    /**
+     * appended message tag define
```
Comment (vongosling) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j/RocketmqLog4jAppender.java]: Capitalize the first letter~

File: style/rmq_checkstyle.xml [hunk 10]
```diff
@@ -30,6 +30,12 @@
     <!-- header -->
     <module name="RegexpHeader">
         <property name="header" value="/\*\nLicensed to the Apache Software Foundation*"/>
+        <property name="fileExtensions" value="java"/>
+    </module>
+
+    <module name="RegexpHeader">
```
Comment (vongosling) [inline style/rmq_checkstyle.xml]: Why change this rule~

Comment (lindzh) [inline style/rmq_checkstyle.xml]: Old check style is not supported When we add a file such as properties config file.

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j/RocketmqLog4jAppender.java (hunk 9 above)
Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j/RocketmqLog4jAppender.java]: ok

**Comment by lizhanhui** [comment]
Hi, @lindzh could you further justify this PR? Aka, explain why it's reasonable to create a custom appender?

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+import org.apache.rocketmq.client.producer.MQProducer;
+
+import java.io.Serializable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Log4j2 Appender Component
+ */
+@Plugin(
+        name = "Rocketmq",
```
Comment (vongosling) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java]: Apache RocketMQ

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+        category = "Core",
```
Comment (vongosling) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java]: Core or core ?

File: logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+import ch.qos.logback.core.AppenderBase;
+import ch.qos.logback.core.Layout;
+import ch.qos.logback.core.spi.PreSerializationTransformer;
+import ch.qos.logback.core.status.ErrorStatus;
+import org.apache.rocketmq.common.message.Message;
+import org.apache.rocketmq.logappender.common.ProducerInstance;
+import org.apache.rocketmq.client.producer.MQProducer;
+
+/**
+ * Logack Appender Component
```
Comment (vongosling) [inline logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java]: Logback

File: logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    private String tag;
+
+    /**
+     * Whitch topic to send log messages
+     */
+    private String topic;
+
+    /**
+     * Rokcetmq nameserver address
```
Comment (vongosling) [inline logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java]: RocketMQ

File: logappender/src/test/java/org/apache/rocketmq/logappender/AbstractTestCase.java [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        Start nameserver
+         */
+        startNamesrv();
+
+        /*
+        Start broker
+         */
+        startBroker();
+
+        Thread.sleep(2000);
```
Comment (vongosling) [inline logappender/src/test/java/org/apache/rocketmq/logappender/AbstractTestCase.java]: CountDownLatch

**Comment by lindzh** [comment]
readme has ben removed

File: logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java (hunk 14 above)
Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/logback/RocketmqLogbackAppender.java]: fixed

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java (hunk 12 above)
Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java]: It's Core,write as log4j2 JMSAppender,refer https://github.com/apache/logging-log4j2/blob/dbd2d252a1b4139a9bd9eb213c89f28498db6dcf/log4j-core/src/main/java/org/apache/logging/log4j/core/appender/mom/JmsAppender.java,
has ben fix to constant.

File: logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+            byte[] bytes = this.getLayout().toByteArray(event);
+            Message msg = new Message(topic, tag, bytes);
+            msg.getProperties().put(ProducerInstance.APPENDER_TYPE, ProducerInstance.LOG4J2_APPENDER);
+
+            //Send message and do not wait for the ack from the message broker.
+            producer.sendOneway(msg);
+        } catch (Exception e) {
+            ErrorHandler handler = this.getHandler();
+            if (handler != null) {
+                handler.error("Could not send message in RocketmqLog4j2Appender [" + this.getName() + "].", e);
```
Comment (zhouxinyu) [inline logappender/src/main/java/org/apache/rocketmq/logappender/log4j2/RocketmqLog4j2Appender.java]: May be we want to know which log message isn't send to server?

**Comment by zhouxinyu** [comment]
BTW, do we use the batch feature in this appender?

File: logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        MQProducer p = producerMap.get(genKey);
+        if (p != null) {
+            return p;
+        }
+
+        DefaultMQProducer defaultMQProducer = new DefaultMQProducer(group);
+        defaultMQProducer.setNamesrvAddr(nameServerAddress);
+        MQProducer beforeProducer = null;
+        //cas put producer
+        synchronized (producerMap) {
```
Comment (zhouxinyu) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]: It seems that there is no need to use `synchronized `?

File: logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+    public static MQProducer getInstance(String nameServerAddress, String group) throws MQClientException {
+        if (group == null) {
+            group = DEFAULT_GROUP;
+        }
+
+        String genKey = genKey(nameServerAddress, group);
+        MQProducer p = producerMap.get(genKey);
+        if (p != null) {
+            return p;
```
Comment (zhouxinyu) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]: Maybe `p` isn't a started producer instance momently?

File: logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        return defaultMQProducer;
+    }
+
+
+    public static void removeAndClose(String nameServerAddress, String group) {
+        if (group == null) {
+            group = DEFAULT_GROUP;
+        }
+        String genKey = genKey(nameServerAddress, group);
+        MQProducer producer = producerMap.get(genKey);
```
Comment (zhouxinyu) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]:
```
MQProducer producer = producerMap.get(genKey);
producerMap.remove(genKey);
```
can be replaced by:

```
MQProducer producer = producerMap.remove(genKey);
```

Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]: Yes,makes it more simple.

File: logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java (hunk 17 above)
Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]: As log initlizer is serializable,there is no need to user synchronized

File: logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java (hunk 18 above)
Comment (lindzh) [inline logappender/src/main/java/org/apache/rocketmq/logappender/common/ProducerInstance.java]: As log initlizer is serializable,if it has ben started before,it must be started.

**Comment by dongeforever** [comment]
@lindzh this PR is merged. You may close this PR now

**Pull request closed without merging**